### PR TITLE
attr.attrs can take a list of `converters`

### PR DESCRIPTION
**Pull request by ashb**
attrs v20.1.0 added support for taking a list of converter objects
https://www.attrs.org/en/stable/changelog.html#id45 (PR 618)

### CODE REVIEW

File: stubs/attrs/attr/__init__.pyi [hunk 1]
```diff
@@ -49,7 +49,7 @@ class Attribute(Generic[_T]):
     order: bool
     hash: Optional[bool]
     init: bool
-    converter: Optional[_ConverterType[_T]]
+    converter: Optional[Union[_ConverterType[_T],List[_ConverterType[_T]],Tuple[_ConverterType[_T]]]
```
Comment (ashb) [inline stubs/attrs/attr/__init__.pyi:52]:
```suggestion
    converter: Optional[Union[_ConverterType[_T],List[_ConverterType[_T]],Tuple[_ConverterType[_T]]]]
```

File: stubs/attrs/attr/__init__.pyi [hunk 2]
```diff
@@ -105,7 +105,7 @@ def attrib(
     init: bool = ...,
     metadata: Optional[Mapping[Any, Any]] = ...,
     type: Optional[Type[_T]] = ...,
-    converter: Optional[_ConverterType[_T]] = ...,
+    converter: Optional[Union[_ConverterType[_T],List[_ConverterType[_T]],Tuple[_ConverterType[_T]]] = ...,
```
Comment (ashb) [inline stubs/attrs/attr/__init__.pyi:108]:
```suggestion
    converter: Optional[Union[_ConverterType[_T],List[_ConverterType[_T]],Tuple[_ConverterType[_T]]]] = ...,
```

File: stubs/attrs/attr/__init__.pyi [hunk 3]
```diff
@@ -123,7 +123,7 @@ def attrib(
...
```
Comment (ashb) [inline stubs/attrs/attr/__init__.pyi:126]:
```suggestion
    converter: Optional[Union[_ConverterType[_T],List[_ConverterType[_T]],Tuple[_ConverterType[_T]]]] = ...,
```

File: stubs/attrs/attr/__init__.pyi [hunk 4]
```diff
@@ -141,7 +141,7 @@ def attrib(
     init: bool = ...,
     metadata: Optional[Mapping[Any, Any]] = ...,
     type: object = ...,
-    converter: Optional[_ConverterType[_T]] = ...,
+    converter: Optional[Union[_ConverterType[_T],List[_ConverterType[_T]],Tuple[_ConverterType[_T]]] = ...,
```
Comment (ashb) [inline stubs/attrs/attr/__init__.pyi:144]:
```suggestion
    converter: Optional[Union[_ConverterType[_T],List[_ConverterType[_T]],Tuple[_ConverterType[_T]]]] = ...,
```

**Review by ashb — COMMENTED** [review]
The perils of editing in a webbrowser

**Comment by ashb** [comment]
Documented here https://www.attrs.org/en/stable/api.html#attr.converters.pipe

File: stubs/attrs/attr/__init__.pyi [hunk 5]
```diff
@@ -49,7 +49,7 @@ class Attribute(Generic[_T]):
     order: bool
     hash: Optional[bool]
     init: bool
-    converter: Optional[_ConverterType[_T]]
+    converter: Optional[Union[_ConverterType[_T],List[_ConverterType[_T]],Tuple[_ConverterType[_T]]]]
```
Comment (Akuli) [inline stubs/attrs/attr/__init__.pyi:52]: `Tuple[Foo]` is a tuple that contains exactly one element, of type `Foo`. You probably meant `Tuple[Foo, ...]` instead.

Smaller nits:
- You will need to run `black`. See README.
- The `|` operator for unions is now preferred, so `Optional[Union[Foo, Bar]]` can be written as `Foo | Bar | None`.

Comment (ashb) [reply stubs/attrs/attr/__init__.pyi:52]: Yeah, black failures were cos I edited this via github. I'll do it properly next time :)

On `|` for Union -- I'm too used to targeting Py 3.6 that it's ingrained that I can't use that.

If we can use `|` (which is py 3.10 ISTR) then canwe also use `list[]` and `tuple[]`?

Comment (srittau) [reply stubs/attrs/attr/__init__.pyi:52]: Yes, `list` and `tuple` are also preferred, **but** `tuple[X, ...]` doesn't work yet. Also, nothing of this works in type aliases yet.

Comment (ashb) [reply stubs/attrs/attr/__init__.pyi]: Updated, I didn't change the other uses of `Optional[]` in this file as it felt to me scope creep of this PR, but I can do if you'd like.

**Comment by Akuli** [comment]
When looking at attrs source, I noticed that they already have type hints in the source and a `py.typed` file. So the stubs in typeshed are used only with older versions of attr (according to changelog, before 18.2.0). Seems like older attrs also support list of converters and don't need this PR?

**Comment by ashb** [comment]
Hmmm curious, I missed that.

This may be a bug in Pyright where it is favouring the version from typeshed.

I'll have to check again and see if mypy also complains here.

**Comment by srittau** [comment]
Thank you for your effort! I am going to close this, since we will remove attrs (#5581), which will hopefully prevent confusion in the future.

**Pull request closed without merging**